### PR TITLE
add way to populate `next_level` property in storyNode`

### PR DESCRIPTION
**Pull request by tongxiang**
#### What's this PR do?

Adds functionality to the exporter--provides a way for the `next_level` property of `END-LEVELX-GROUP` to be populated. 

Note that all of the `_build` functions have been refactored to take in a `config` argument--we pass the config object from one build function to the next because the end-level and end-game build functions require data from other passages. (For instance, `_buildEndLevels` requires data produced by `_buildRegularLevels`, and `_buildEndGame` will require data produced by `_buildEndLevels`.)
#### How should this be manually tested?

Exporting an sms game with multiple levels set up in twine. 
#### Questions:

Is there a better way to execute the logic on lines 386-390?

### CODE REVIEW

File: js/exportsms.js [hunk 1]
```diff
@@ -215,13 +215,19 @@ var exportsms = function() {
         j,
         k;
 
-    partialStory = {};
+
+    if (config.story) {
+      partialStory = config.story;
+    } else {
+      partialStory = {}
```
Comment (jonuy) [inline js/exportsms.js]: semicolon

File: js/exportsms.js [hunk 2]
```diff
@@ -293,10 +304,18 @@ var exportsms = function() {
...
+        storyNode,
+        keyArray,
+        testKey
         ;
 
-    partialStory = {};
+    if (config.story) {
+      partialStory = config.story;
+    } else {
+      partialStory = {}
```
Comment (jonuy) [inline js/exportsms.js]: semicolon. + perhaps putting the `else {` on a new line.

File: js/exportsms.js [hunk 3]
```diff
@@ -355,13 +374,34 @@ var exportsms = function() {
           choice = configObject.choices[j];
           if (endLevelGroupSuccessFailureStatus == choice.flag.toUpperCase() && !choice.next) {
             choice.next = parseInt(passageDatum.optinpath, 10);
-            partialStory[nameString] = configObject;
             break;
           }
         }
+
+        nextLevelNumber = parseInt(levelNumber, 10) + 1
```
Comment (jonuy) [inline js/exportsms.js]: semicolon

File: js/exportsms.js [hunk 4]
```diff
@@ -355,13 +374,34 @@ var exportsms = function() {
...
+        if (nextLevelNumber <= TOTAL_NUMBER_OF_LEVELS) {
+          var nextLevelString = 'L' + nextLevelNumber + '0'
```
Comment (jonuy) [inline js/exportsms.js]: semicolon and maybe declare `var nextLevelString` at the top of the function with the rest of the vars.

File: js/exportsms.js [hunk 5]
```diff
@@ -355,13 +374,34 @@ var exportsms = function() {
...
+          for (var storyNode in config.story) {
```
Comment (jonuy) [inline js/exportsms.js]: `storyNode` is already declared at the top of this function. Don't need the `var` here.

File: js/exportsms.js [hunk 6]
```diff
@@ -355,13 +374,34 @@ var exportsms = function() {
...
+                testKey = keyArray[k];
+                if (config.story[testKey].key == nextLevelString) {
+                  configObject.next_level = keyArray[k];
+                  break;
+                }
+              }
+            }
+          }
+        } else {
+          configObject.next_level = 'END-GAME'
```
Comment (jonuy) [inline js/exportsms.js]: semicolon

File: js/exportsms.js [hunk 7]
```diff
@@ -355,13 +374,34 @@ var exportsms = function() {
...
             break;
           }
         }
+
+        nextLevelNumber = parseInt(levelNumber, 10) + 1
+        if (nextLevelNumber <= TOTAL_NUMBER_OF_LEVELS) {
+          var nextLevelString = 'L' + nextLevelNumber + '0'
+          for (var storyNode in config.story) {
+            if (config.story.hasOwnProperty(storyNode) && config.story[storyNode].key == nextLevelString) {
+              keyArray = Object.keys(config.story); // This feels like a very hacky way to retrieve the key, but might be the only way.
```
Comment (jonuy) [inline js/exportsms.js]: I think this is fine

Comment (tongxiang) [inline js/exportsms.js]: add comment about why we added `.hasOwnProperty`

**Comment by jonuy** [comment]
Other than my semicolon crusade, this seems good.

The main concern I have now is how you're capping the max number of levels here in code at 6. It's not immediately apparent to the user that this is a thing, and something I could foresee us changing in the future. Is there a property or text that the user can add instead that would explicitly connect this to `END-GAME`?

**Comment by tongxiang** [comment]
Yeah, that seems like a great idea. And perhaps something to include in the high-level config passage object? 

This seems like an issue for another PR--I'll put in an issue now.

**Comment by jonuy** [comment]
Could be a high-level config passage. Could be something that's somehow auto-detected? Like, if no next level is found, just assume it should go to END-GAME?

**Comment by tongxiang** [comment]
Oh, right. That would be the best way. I'll implement it that way.

**Pull request merged**